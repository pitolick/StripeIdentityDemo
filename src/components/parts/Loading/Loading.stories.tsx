import React from 'react'
import { Meta, StoryObj } from '@storybook/react'
import { Loading } from './Loading'

const meta: Meta<typeof Loading> = {
  title: 'parts/Loading',
  component: Loading,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Loading>

/**
 * デフォルト
 */
export const Default: Story = {
  args: {
    show: false,
  },
}

/**
 * 子要素
 */
export const WithChildren: Story = {
  args: {
    show: false,
    children: <div className="text-white">Loading...</div>,
  },
}
