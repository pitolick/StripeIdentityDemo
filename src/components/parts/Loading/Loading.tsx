/**
 * ローディング
 * @module Loading
 * @todo ReferenceError: document is not definedが出ることがある？
 */
'use client'
import React, { useEffect, useState } from 'react'
import { createPortal } from 'react-dom'
import { Spinner } from '@/components/elements'

/**
 * ローディングのプロパティ
 */
export interface LoadingProps {
  /**
   * 表示フラグ
   */
  show: boolean
  /**
   * 子要素
   */
  children?: React.ReactNode
}

/**
 * ローディングコンポーネント
 */
export const Loading: React.FC<LoadingProps> = ({ show, children }) => {
  // アニメーションフラグ
  const [animation, setAnimation] = useState(show)

  // showが変わってから0.2秒後にアニメーションを開始する
  useEffect(() => {
    if (!show) {
      setTimeout(() => {
        setAnimation(show)
      }, 200)
    } else {
      setAnimation(show)
    }
  }, [show])

  // ドキュメントがマウントされたら表示する
  const [documentMouned, setDocumentMounted] = useState(false)
  useEffect(() => {
    setDocumentMounted(true)
  }, [])

  return (
    <>
      {documentMouned &&
        createPortal(
          <div className={`fixed inset-0 z-50 bg-gray-900/50 transition duration-200 ${show ? 'opacity-100' : 'opacity-0'} ${!animation && 'hidden'}`}>
            <div className="flex size-full flex-col items-center justify-center">
              <Spinner size="lg" />
              {children}
            </div>
          </div>,
          document.body,
        )}
    </>
  )
}
