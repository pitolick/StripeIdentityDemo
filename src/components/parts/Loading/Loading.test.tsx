import React from 'react'
import { render, act } from '@testing-library/react'
import { Loading } from './Loading'

describe('Loadingコンポーネント', () => {
  it('表示フラグがtrueの場合、ローディングが表示されること', () => {
    act(() => {
      const { container } = render(<Loading show={true} />)
      expect(container).toMatchSnapshot()
    })
  })

  it('表示フラグがfalseの場合、ローディングが非表示になること', () => {
    act(() => {
      const { container } = render(<Loading show={false} />)
      expect(container).toMatchSnapshot()
    })
  })

  it('表示フラグがfalseからtrueに変わった場合、0.2秒待ってからローディングが表示されること', () => {
    act(() => {
      jest.useFakeTimers()
      const { container, rerender } = render(<Loading show={false} />)
      rerender(<Loading show={true} />)
      expect(container).toMatchSnapshot()
      jest.advanceTimersByTime(200)
      expect(container).toMatchSnapshot()
    })
  })
})
