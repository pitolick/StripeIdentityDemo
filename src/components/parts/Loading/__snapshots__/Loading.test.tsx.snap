// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loadingコンポーネント 表示フラグがfalseからtrueに変わった場合、0.2秒待ってからローディングが表示されること 1`] = `<div />`;

exports[`Loadingコンポーネント 表示フラグがfalseからtrueに変わった場合、0.2秒待ってからローディングが表示されること 2`] = `<div />`;

exports[`Loadingコンポーネント 表示フラグがfalseの場合、ローディングが非表示になること 1`] = `<div />`;

exports[`Loadingコンポーネント 表示フラグがtrueの場合、ローディングが表示されること 1`] = `<div />`;
